numpy/_core/src/multiarray/usertypes.c

```diff
@@ -145,32 +145,42 @@ PyArray_InitArrFuncs(PyArray_ArrFuncs *f)
   defined in arraytypes.inc
 */
 /*NUMPY_API
-  Register Data type
-  Does not change the reference count of descr
+ * Register Data type
+ *
+ * Creates a new descripto from a prototype one.  The prototype is not
+ * a valid descriptor on 2.x, but is used as the prototype instance on 1.x.
+ * Code must use `descr = PyArray_DescrFromType(num);` after successful
+ * registration.
+ *
+ * This function copies all internal references on 2.x only.  This should be
+ * irrelevant, since any internal reference is immortal.
 */
 NPY_NO_EXPORT int
-PyArray_RegisterDataType(PyArray_Descr *descr)
+PyArray_RegisterDataType(PyArray_DescrProto *descr_proto)
 {
     PyArray_Descr *descr2;
     int typenum;
     int i;
     PyArray_ArrFuncs *f;
 
-    /* See if this type is already registered */
+    /* 
+     * See if this type is already registered (assumes type_num is initialized)
+     * to 0 or -1.
```

Comment:
Assumes which typenum is initialized? "(assumes descr_proto typenum is initialized to 0 or -1)"